### ISSUE TITLE
React-projects with CRA & typescript can't use JSX with ts-ls

### ISSUE THREAD

**Issue by adimit** [emacs-lsp/lsp-mode#2629]
**Describe the bug**
In a fresh CRA-created typescript React App, JSX tags are not recognized. The error message is `Cannot use JSX unless the --jsx flag is provided`. This *may* be related to #2374. Workaround: setting `"jsx": "preserve"` in `tsconfig.json` works around the issue, *however*, due to recent changes in React 17, new CRA-projects need that setting on `"jsx": "react-jsx"`.

**To Reproduce**
I'm using this load script for LSP:
```emacs-lisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)
(package-refresh-contents)
(package-install 'lsp-mode)
(package-install 'web-mode)
(add-to-list 'auto-mode-alist '("\\.tsx$". web-mode))
(add-hook 'web-mode-hook #'lsp)
```
Save as `load-lsp-mode.el`.

```shell
yarn create react-app test-app-lsp-mode --template typescript
cd test-app-lsp-mode
env HOME=$(pwd) emacs -Q --load path/to/load-lsp-mode.el --file src/App.tsx
```

**Expected behavior**
JSX tags should work without a problem.

Now, I *do* think this may be a problem with `lsp-mode` *specifically*, because the same works with `eglot`. Note that you have to install `typescript-language-server` globally with `eglot` (i.e. it has to be in your path.) I use the following load file:

```emacs-lisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)
(package-refresh-contents)
(package-install 'web-mode)
(package-install 'project)
(package-install 'eglot)
(setq eglot-server-programs nil)
(add-to-list 'eglot-server-programs `(web-mode . ("typescript-language-server" "--stdio")))
(add-hook 'web-mode-hook 'eglot-ensure)
(add-to-list 'auto-mode-alist '("\\.tsx$". web-mode))
```
Save as `load-eglot.el`

```shell
yarn global add typescript-language-server # Skip this if typescript-language-server is available in your path
yarn create react-app test-app-eglot --template typescript
cd test-app-eglot
env HOME=$(pwd) emacs -Q --load path/to/load-eglot.el --file src/App.tsx
```

This same behaviour is not exhibited there, and JSX tags work fine.

**Which Language Server did you use**
`ts-ls`

Emacs version 27.1.9, lsp-mode is latest melpa, `20210219.1128`. `create-react-app@4.0.2`

**OS**
Ubuntu 20.10

I'd love to provide further info, but I don't know where to start debugging this, as I'm quite ignorant about how `lsp-mode` interacts with lsp-servers. Thank you for any help on this issue!

**Comment by yyoncho** [comment]
The issue is caused by the fact that lsp-mode uses the tsserver that is automatically installed lsp-mode facilities while in the other case when there is no tsserver passed and the typescript language server picks the tsserver that is part of the project. Probably they are different versions.

**Comment by adimit** [comment]
Good idea, but I did make sure the versions of ts-ls were the same:

```
$ typescript-language-server --version
0.5.1
$ .emacs.d/.cache/lsp/npm/typescript-language-server/bin/typescript-language-server --version
0.5.1
```

**Comment by yyoncho** [comment]
`typescript-language-server` is only wrapper over `tsserver`. You may grep which tsserver is started in both cases. I already have an idea of how to fix the issue and at the same time keep the OOTB experience.

**Comment by adimit** [comment]
I started the above `eglot` load file with the following line for `eglot-server-programs`:

```
(add-to-list 'eglot-server-programs `(web-mode . ("typescript-language-server" "--stdio" "--tsserver-path" ,(expand-file-name ".emacs.d/.cache/lsp/npm/typescript/bin/tsserver"))))
```

This still works (with `eglot`), and should use the same underlying `tsserver`.

**Comment by yyoncho** [comment]
Let me prepare a PR. When I am testing with lsp-mode I see the error when we explicitly pass the tsserver as param but I don't see it when lsp-mode does not pass that path.

**Comment by adimit** [comment]
OK, thanks a lot for your fast response! I actually further adapted the line above to:

```
(setq eglot-server-programs nil)
(add-to-list
 'eglot-server-programs
 `(web-mode . (,(expand-file-name ".emacs.d/.cache/lsp/npm/typescript-language-server/bin/typescript-language-server")
               "--stdio"
               "--tsserver-path"
               ,(expand-file-name ".emacs.d/.cache/lsp/npm/typescript/bin/tsserver"))))
```

Which means `eglot` definitely uses the same versions of both `ts-ls` and `tsserver`. It still works there (though there is another issue in that it issues spurious warnings :thinking:, which `lsp-mode` doesn't) but not in `lsp-mode`.

**Comment by yyoncho** [comment]
Proposed fix at #2633

**Comment by adimit** [comment]
Woot! Tried it out and it fixes the problem!

**Comment by adimit** [comment]
Oh wait, sorry, I don't think it worked. I'm using your `tsserver` branch; I think that I had the workaround (fix in `tsconfig`) still enabled in the repository I originally saw the issue with and that's why it seemed to work. I've now tried it again with the new branch and can still see the bug.

**Comment by yyoncho** [comment]
Hm, so on my side it seems like it works even without the PR but before that, I was able to reproduce the issue without the fix. I did several things:
1. Uninstalled global typescript
2. Moved from node 12 to node 10. 

At this point, I am unable to reproduce the issue and it seems like there something in the environment that triggers the issue.

**Comment by yyoncho** [comment]
I am testing with this:

```
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)
(package-refresh-contents)
(package-install 'web-mode)
(package-install 'lsp-mode)
(package-install 'flycheck)
(add-hook 'web-mode-hook 'lsp)
(add-to-list 'auto-mode-alist '("\\.tsx$". web-mode))
```
and then:

`env HOME=$(pwd) emacs -Q --load ../load-lsp-mode.el --file src/App.tsx`

(which will install everything + the server).

**Comment by adimit** [comment]
Interesting! I shall try to reproduce the issue in a docker environment, but I won't get around to that before the evening. My node version is 12.19.0

**Comment by yyoncho** [comment]
Docker recipe will be fine. Do you reproduce the issue with the config from the previous reply?

**Comment by adimit** [comment]
> Docker recipe will be fine. Do you reproduce the issue with the config from the previous reply?

Yes, this is what I did:
```
yarn create react-app lsp-mode-test --template typescript
cd lsp-mode-test
env HOME=(pwd) emacs -Q --load ../load-lsp-mode.el --file src/App.tsx
``` 

With the contents of `load-lsp-mode.el` copied from your file in [this comment](https://github.com/emacs-lsp/lsp-mode/issues/2629#issuecomment-782683953). The result is in this screenshot (error message in the mode line.) 
![emacs](https://user-images.githubusercontent.com/15893/108604919-1abe0b00-73b1-11eb-8302-0c4935abe7f8.png)

----

So I tried the docker repro like this:

```
$ docker run --rm -ti --entrypoint "/bin/bash" node:12-buster
# inside the container:
$ apt update && apt install emacs
$ yarn  create react-app --template typescript test
$ cd test
$ cat << EOF > load.el(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
; Buster only provides Emacs 26, so we need to work around the tls issue
(setq gnutls-algorithm-priority "NORMAL:-VERS-TLS1.3")
(package-initialize)
(package-refresh-contents)
(package-install 'web-mode)
(package-install 'lsp-mode)
(package-install 'flycheck)
(add-hook 'web-mode-hook 'lsp)
(add-to-list 'auto-mode-alist '("\.tsx$". web-mode))
EOF
$ env HOME=$(pwd) emacs -Q --load load.el --file src/App.tsx
```

And it worked fine! It seems something is indeed the case with my environment then. I'll dig around to see what the issue is. Perhaps some globally installed thing, or maybe even the Emacs version (26 on docker, I'm using my own compiled version of Emacs 27. Maybe I'm lacking a build flag…)

**Comment by yyoncho** [comment]
I believe that it is not related to emacs/lsp-mode. It is related to nodejs/tsserver/typescript-language-server. 

> My node version is 12.19.0

I changed my version from 12 to 10 due to an issue with create reactapp so that might be the key. Then the error disappeared and I thought that it is caused by my PR.

**Comment by adimit** [comment]
You're probably right. I removed the binaries for a ts-version I had lying around in `$HOME/.local/bin/`. They were probably old, and perhaps didn't know about the `"jsx": "react-jsx"` setting. This is probably what caused the issue persistently.

Good thing I attempted the docker repro, that put me on the right track. Thanks a lot for helping me debug this issue. It's fixed on my end, so I'll close this.
